In [1]:
# 라이브러리 사용
import pandas as pd
import tensorflow as tf

In [3]:
# 1.과거의 데이터를 준비합니다.
파일경로 = './csv/iris.csv'
아이리스 = pd.read_csv(파일경로)
아이리스.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [4]:
# 원핫인코딩
인코딩 = pd.get_dummies(아이리스)
인코딩.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2       True          False         False
1   4.9  3.0    1.4   0.2       True          False         False
2   4.7  3.2    1.3   0.2       True          False         False
3   4.6  3.1    1.5   0.2       True          False         False
4   5.0  3.6    1.4   0.2       True          False         False

In [5]:
# 칼럼이름 출력
print(인코딩.columns)

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')


In [8]:
# 독립변수, 종속변수
독립 = 인코딩[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속 = 인코딩[['품종_setosa', '품종_versicolor', '품종_virginica']]
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


In [10]:
# 2. 모델의 구조를 만듭니다
X = tf.keras.layers.Input(shape=[4]) # 독립변수 13개 이므로 13
Y = tf.keras.layers.Dense(3, activation='softmax')(X)  # activation='softmax' : 분류 모델중 softmax 사용하겠다.
model = tf.keras.models.Model(X, Y)
# 다중분류손실함수 : categorical_crossentropy는 데이터의 라벨이 원핫인코딩 된 형태의 경우 사용. 
# 보통 softmax 활성화함수 다음에 연계되어 사용되므로 softmax loss로도 불린다.
# sparse_categorical_crossentropy는 데이터의 라벨이 Int형 일 때 사용한다.
model.compile(loss='categorical_crossentropy',  # loss: 손실 함수를 설정합니다. 모델이 예측한 값과 실제 값 사이의 차이를 계산하는 함수입니다.
              metrics = ['accuracy'])
# metrics: 평가 지표를 설정합니다.학습 중 모델의 성능을 평가할 때 사용됩니다. 
# 정확도(Accuracy), 정밀도(Precision), 재현율(Recall), F1 점수 등이 있습니다.

In [13]:
# 3.데이터로 모델을 학습(FIT)합니다.
model.fit(독립, 종속, epochs=100)

Epoch 1/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8522 - loss: 0.4810 
Epoch 2/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 998us/step - accuracy: 0.8401 - loss: 0.4779
Epoch 3/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8482 - loss: 0.4967 
Epoch 4/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 998us/step - accuracy: 0.8321 - loss: 0.4811
Epoch 5/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8765 - loss: 0.4533 
Epoch 6/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 998us/step - accuracy: 0.8608 - loss: 0.4627
Epoch 7/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8639 - loss: 0.4650 
Epoch 8/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8695 - loss: 0.4661 
Epoch 9/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8308 - loss: 0.4613 
Epoch 10/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 998us/step - accuracy: 0.8965 - loss: 0.4472
Epoch 11/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8613 - loss: 0.4783 
Epoch 12/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 755us/step - accuracy: 0.8591

In [14]:
# 모델을 이용합니다.
# 맨 처음 데이터 5개
print(model.predict(독립[:5]))
print(종속[:5])  # 실제 정답

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
[[0.9508982  0.04139617 0.00770557]
 [0.8989405  0.08651546 0.01454408]
 [0.93557125 0.05380743 0.01062138]
 [0.9145704  0.07116392 0.0142656 ]
 [0.95951635 0.03367552 0.00680805]]
   품종_setosa  품종_versicolor  품종_virginica
0       True          False         False
1       True          False         False
2       True          False         False
3       True          False         False
4       True          False         False


In [15]:
# 맨 끝 데이터 5개
print(model.predict(독립[-5:]))
print(종속[-5:])  # 실제 정답

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step
[[0.002299   0.2612445  0.73645645]
 [0.00381609 0.40990138 0.58628255]
 [0.00516657 0.3450401  0.6497933 ]
 [0.00355681 0.19392669 0.8025165 ]
 [0.01005006 0.3583298  0.63162005]]
     품종_setosa  품종_versicolor  품종_virginica
145      False          False          True
146      False          False          True
147      False          False          True
148      False          False          True
149      False          False          True


In [16]:
# weights & bias 출력. 학습한 가중치
print(model.get_weights())

[array([[-0.35968933,  0.04584568, -0.3984503 ],
       [ 1.2256223 , -0.52324146, -0.1417371 ],
       [-0.3401752 ,  0.49954578,  0.5451911 ],
       [-1.589324  , -0.08790711,  1.5529562 ]], dtype=float32), array([ 0.8977075 ,  0.34039077, -0.802287  ], dtype=float32)]


첫번째 컬럼은 바로 setosa를 예측하는 수식이라고 보시면 됩니다. 첫번째 -0.35968933은 꽃잎의 길이와 곱해주는 가중치이고, 두번째1.2256223은 꽃잎의 폭에 곱해주는 가중치, 그리고 세번째  -0.3401752 는 꽃받침의 길이를 곱해주는 가중치, 네번째 인 -1.589324 은 꽃받침의 폭을 곱해주는 가중치입니다. 그래서 이렇게 수식을 하나 만들고 마지막에 bias 인 0.8977075을 더해주면 setosa인지 아닌지를 가늠할 수 있는 값이 나오는 것입니다. 여기에 softmax를 한번 씌워주면 그 확률이 나와서 아까 보셨듯이 0.8***뭐 이렇게 나왔던 거죠. 그리고 두번째 것은 versicolor인째 컬럼이 virginica일 확률에 대해서 계산해주는 수식에 이렇게 가중치를 나타내는 것입니다. 여러분들이 이렇게 종속변수가 범주형일때, 분류 모델을 만드는 법을 학습을 했습니다. 그리고 이런 범주형 데이터를 학습시키기 위해서는 수치형으로 변형이 필요한데 그변형시켜 주는 방법까지 학습이 끝났습니다.